File: app/api/categories/route.ts
// app/api/categories/route.ts
import { createClient } from "@supabase/supabase-js";
import { NextResponse } from "next/server";

export const dynamic = "force-dynamic";
export const revalidate = 0;

const url = process.env.NEXT_PUBLIC_SUPABASE_URL!;
const key = process.env.SUPABASE_SERVICE_ROLE_KEY || process.env.NEXT_PUBLIC_SUPABASE_ANON_KEY!;

export async function GET() {
  try {
    if (!url || !key) {
      return NextResponse.json({ error: "Supabase env ontbreekt" }, { status: 500, headers: { "X-Category-Source": "error-env" } });
    }
    const sb = createClient(url, key, { auth: { persistSession: false } });

    const { data, error } = await sb
      .from("categories")
      .select("id,name,parent_id,position")
      .order("position", { ascending: true });

    if (error) {
      return NextResponse.json({ error: error.message }, { status: 500, headers: { "X-Category-Source": "error-db" } });
    }

    const rows = data || [];
    if (!rows.length) return new NextResponse(null, { status: 204, headers: { "X-Category-Source": "empty-db" } });

    const parents = rows.filter((r) => !r.parent_id);
    const byParent = new Map<string, { id: string; name: string }[]>();
    for (const r of rows) {
      if (r.parent_id) {
        const arr = byParent.get(r.parent_id) || [];
        arr.push({ id: r.id, name: r.name });
        byParent.set(r.parent_id, arr);
      }
    }

    const shaped = parents.map((p) => ({
      id: p.id,
      name: p.name,
      subcategories: (byParent.get(p.id) || []).map((c) => c.name),
    }));

    if (!shaped.length) return new NextResponse(null, { status: 204, headers: { "X-Category-Source": "no-parents" } });
    return NextResponse.json(shaped, { status: 200, headers: { "X-Category-Source": "supabase" } });
  } catch (e: unknown) {
    const errorMessage = typeof e === "object" && e !== null && "message" in e ? (e as { message?: string }).message : "Onbekend";
    return NextResponse.json({ error: errorMessage || "Onbekend" }, { status: 500, headers: { "X-Category-Source": "exception" } });
  }
}

File: app/api/listings/route.rs
// app/api/listings/route.ts
import { NextResponse } from "next/server";
import { createClient } from "@supabase/supabase-js";

export const dynamic = "force-dynamic";
export const revalidate = 0;

const url = process.env.NEXT_PUBLIC_SUPABASE_URL!;
const key =
  process.env.SUPABASE_SERVICE_ROLE_KEY || process.env.NEXT_PUBLIC_SUPABASE_ANON_KEY!;

export async function GET(req: Request) {
  try {
    if (!url || !key) {
      return NextResponse.json({ error: "Supabase env ontbreekt" }, { status: 500 });
    }
    const supabase = createClient(url, key, { auth: { persistSession: false } });

    const { searchParams } = new URL(req.url);
    const category = searchParams.get("category") || "";
    const subcategory = searchParams.get("subcategory") || "";
    const page = Number(searchParams.get("page") || "1");
    const pageSize = Math.min(50, Math.max(1, Number(searchParams.get("pageSize") || "24")));
    const from = (page - 1) * pageSize;
    const to = from + pageSize - 1;

    let query = supabase
      .from("listings")
      .select("id,title,price,images,main_photo,categories,location,state,created_at", { count: "exact" })
      .order("created_at", { ascending: false })
      .range(from, to);

    // Filter: category aanwezig in JSONB array
    if (category) {
      query = query.contains("categories", [category] as any);
    }
    // Subcategorie als tweede element (ook gewoon "contains" is oké)
    if (subcategory) {
      query = query.contains("categories", [category, subcategory] as any);
    }

    const { data, error, count } = await query;
    if (error) throw error;

    return NextResponse.json({
      page,
      pageSize,
      total: count ?? 0,
      items: data ?? [],
    });
  } catch (e: any) {
    return NextResponse.json({ error: e?.message || "Unknown" }, { status: 500 });
  }
}
